### PR TITLE
[3/3] no-std support phase I

### PR DESCRIPTION
**Pull request by japaric**
this PR implements phase I of RFC #1399 . NOTE that this PR is built on top of PR #1583 

originally phase I aimed to only allowing client functionality in no-std context but turns out that the changes were sufficient to also enable server functionality as evidenced in #1534 

breaking changes: implementing the RFC requires no breaking changes

blockers:
- [x] #1583 

this PR is best reviewed on a commit by commit basis (do skip PR#1583 commits)

### CODE REVIEW

**Comment by ctz** [comment]
> * `pki_types::UnixTime` will need to gain a constructor so that `no_std` users can implement the `GetCurrentTime` trait

I think this exists and is `UnixTime::since_unix_epoch(core::time::Duration)`

**Comment by djc** [comment]
I think this would be much easier to digest if it was split into smaller commits, for example:

* One commit for importing `Vec` explicitly
* One commit for `TimeProvider`
* One commit for dealing with the cache
* One commit for compiling out all the server stuff

**Comment by japaric** [comment]
the PR has been split in commits to make it easier to review. all the "rm (..)" commits can be squashed at a later time. also, there are several warning that I haven't fixed and they involve cfg-ing away more code. for now, I have only removed what's need to make rustls compile with `#![no_std]`

**Comment by djc** [comment]
I think the `rm *` commits would actually be nice to keep, assuming each can pass `cargo check` on its own? Would make it a lot easier to review!

**Comment by japaric** [comment]
> I think the rm * commits would actually be nice to keep, assuming each can pass cargo check on its own?

I reordered the commits a bit and now only these contiguous commits:

- add std feature
- no-std: remove field from `*Error::Other` variants
- no-std: add TimeProvider to ClientConfig
- no-std: rm TicketSwitcher
- no-std: add TimeProvider to ServerConfig

make `cargo b --no-default-features` fail. `cargo b` works on all the commits of this PR.

---

The first commit of this PR basically adds this to `src/lib.rs`

``` rust
#![no_std]
extern crate std;
```

this switches the prelude from `std::prelude` to `core::prelude` and forces one to explicitly import anything that's not in `core`, that is mainly `alloc` API.

I think it would be worthwhile to land that change in a separate PR ahead of this one: it'll prevent relying on implicit `alloc` imports so all future PRs will have to include explicit `alloc` imports to avoid breaking the build. That would not only prevent this PR from bitrotting but I think it would help maintain no-std support long term. I'll submit that commit as a separate PR with a fleshed out rationale.

**Comment by japaric** [comment]
quick update: I have rebased this PR now that #1524 has been merged. I have also pushed a commit that makes `cargo b --no-default-features --features ring` work on Linux. however, I tried to build rustls with that set of features (`-std +ring`) for the `x86_64-unknown-none` target and it failed to compile.

turns out that *ring* has this sealed `SecureRandom` trait (note that there are two of them in the *ring* codebase; one is public and the other is not) that's implemented for `SystemRandom` but only for a [known list of OSes](https://github.com/briansmith/ring/blob/2e8363b433fa3b3962c877d9ed2e9145612f3160/src/rand.rs#L127-L149). as "OS agnostic" targets like `x86_64-unknown-none` have `os = "none"` in their target specification, `SystemRandom` lacks the `SecureRandom` trait and that causes problems in the rustls codebase: [`RsaSigner`](https://github.com/rustls/rustls/blob/ac10422a898335230f64104b169461a50e9d61db/rustls/src/crypto/ring/sign.rs#L148), [`KxGroup`](https://github.com/rustls/rustls/blob/ac10422a898335230f64104b169461a50e9d61db/rustls/src/crypto/ring/kx.rs#L27), etc.

for more context: I'm building a small rustls no-std demo that only depends on `libc` bindings and should work on any OS that has POSIX layer as way to exercise this PR. I'm running the demo on Linux so I do have a "secure random" `getrandom` implementation but I can't signal that to *ring*. I'll explore using an alternative crypto provider as showcased in #1405.

**Comment by djc** [comment]
@japaric probably also good to open a *ring* issue to signal this there and see if there is a better solution there. (Please link the issue here if you do so.)

**Comment by japaric** [comment]
@djc done. submitted briansmith/ring#1734

**Comment by japaric** [comment]
rebased this on top of #1583 and marked it as ready for review

File: rustls/src/client/client_conn.rs [hunk 1]
```diff
@@ -378,7 +419,11 @@ pub enum ServerName {
 
     /// The server is identified by an IP address. SNI is not
     /// done.
+    #[cfg(feature = "std")]
     IpAddress(IpAddr),
+
+    /// Same as the `IpAddress` variant, which is a wrapper around `std::net::IpAddr`, but no-std compatible
+    NoStdIpAddress(webpki::IpAddr),
```
[the remark this answers is not in the document]
Comment (ctz) [reply rustls/src/client/client_conn.rs:426]: I think I'd prefer option (2), awaiting https://doc.rust-lang.org/core/net/index.html to be stabilised.

**Comment by japaric** [comment]
rebased to fix merge conflicts.

thanks djc for the switch to `pki_types::ServerName`. that simplified a few things in this PR.

**Comment by pvdrz** [comment]
rebased on top of https://github.com/rustls/rustls/pull/1583

File: rustls/src/client/tls12.rs [hunk 2]
```diff
@@ -974,6 +977,12 @@ impl ExpectFinished {
             return;
         }
 
+        let now = if let Ok(now) = self.config.get_current_time() {
+            now
+        } else {
+            return;
```
Comment (cpu) [inline rustls/src/client/tls12.rs:983]: Does this early return merit a `debug!` like on L976? I think it would be helpful.

File: rustls/src/client/hs.rs [hunk 3]
```diff
@@ -68,7 +68,9 @@ fn find_session(
             None
         })
         .and_then(|resuming| {
-            let retrieved = persist::Retrieved::new(resuming, UnixTime::now());
+            let now = config.get_current_time().ok()?;
```
Comment (cpu) [inline rustls/src/client/hs.rs:71]: I think matching the err result here and `debug!` logging would be helpful. If the time provider is broken session resumption would silently return `None` for all sessions.

File: rustls/src/time_provider.rs [hunk 4]
```diff
@@ -0,0 +1,47 @@
+//! no-std time provider
+
+use alloc::sync::Arc;
+use core::fmt::Debug;
+
+use pki_types::UnixTime;
+
+/// An object that provides the current time
```
Comment (cpu) [inline rustls/src/time_provider.rs:8]: After the rustdoc headline I think we should mention some of the places where a time provider is used . For certificate validation in particular it's important that the time be accurate or an expired certificate could be accepted. Calling that out explicitly might not hurt.

File: rustls/src/error.rs [hunk 5]
```diff
@@ -523,7 +524,10 @@ impl fmt::Display for Error {
                 write!(f, "the supplied max_fragment_size was too small or large")
             }
             Self::General(ref err) => write!(f, "unexpected error: {}", err),
+            #[cfg(feature = "std")]
             Self::Other(ref err) => write!(f, "other error: {}", err),
+            #[cfg(not(feature = "std"))]
+            Self::Other(..) => f.write_str("write error"),
```
Comment (cpu) [inline rustls/src/error.rs:532]: Is this an appropriate description for all cases or should it say something like "unknown error" or "other unspecified error"?

File: rustls/src/client/client_conn.rs [hunk 6]
```diff
@@ -401,7 +404,13 @@ impl Default for Resumption {
     /// Create an in-memory session store resumption with up to 256 server names, allowing
     /// a TLS 1.2 session to resume with a session id or RFC 5077 ticket.
```
Comment (cpu) [inline rustls/src/client/client_conn.rs:412]: I think this rustdoc comment should be updated to mention that for no-std it actually disables resumption.

**Review by cpu — COMMENTED** [review]
This is great, thank you. I picked up review at 94e507357a31e12030ccc224c21696d910ae475e onward. I only have some small comments.

The number of new `#[cfg(feature = "std")]` flags is a bit dizzying but I trust your process/experience here.

File: rustls/src/client/client_conn.rs (hunk 6 above)
Comment (cpu) [reply rustls/src/client/client_conn.rs:413]: (Though maybe of limited value since this comes back in phase ii)

Comment (pvdrz) [reply rustls/src/client/client_conn.rs:413]: I did it anyway :P

Comment (cpu) [reply rustls/src/client/client_conn.rs:413]: Thanks :-)

File: provider-example/Cargo.toml [hunk 7]
```diff
@@ -18,7 +18,7 @@ p256 = { version = "0.13.2", default-features = false, features = ["alloc", "ecd
 pkcs8 = "0.10.2"
 pki-types = { package = "rustls-pki-types", version = "1" }
 rand_core = { version = "0.6", features = ["getrandom"] }
-rustls = { path = "../rustls", default-features = false, features = ["logging", "tls12"] }
+rustls = { path = "../rustls", default-features = false, features = ["logging", "tls12", "std"] }
```
Comment (djc) [inline provider-example/Cargo.toml:21]: Let's keep features in alphabetic order?

File: rustls/Cargo.toml [hunk 8]
```diff
@@ -20,17 +20,18 @@ aws-lc-rs = { version = "1.5", optional = true }
...
 subtle = { version = "2.5.0", default-features = false }
-webpki = { package = "rustls-webpki", version = "0.102", features = ["std"], default-features = false }
-pki-types = { package = "rustls-pki-types", version = "1", features = ["std"] }
+webpki = { package = "rustls-webpki", version = "0.102", features = ["alloc"], default-features = false }
+pki-types = { package = "rustls-pki-types", version = "1" }
 zeroize = "1.6.0"
 
 [features]
-default = ["logging", "ring", "tls12"]
+default = ["logging", "ring", "tls12", "std"]
```
Comment (djc) [inline rustls/Cargo.toml:28]: Also, here.

File: rustls/src/client/client_conn.rs [hunk 9]
```diff
@@ -326,6 +332,19 @@ impl ClientConfig {
             .copied()
             .find(|skxg| skxg.name() == group)
     }
+
+    pub(super) fn get_current_time(&self) -> Result<UnixTime, Error> {
```
Comment (djc) [inline rustls/src/client/client_conn.rs:345]: Nit: let's kill the `get_()` prefix while we're moving this.

File: rustls/src/client/tls12.rs [hunk 10]
```diff
@@ -974,6 +977,13 @@ impl ExpectFinished {
             return;
         }
 
+        let now = if let Ok(now) = self.config.get_current_time() {
```
Comment (djc) [inline rustls/src/client/tls12.rs:980]: Style: IMO this would be more idiomatic as a `match` expression.

File: rustls/src/time_provider.rs [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+impl TimeProvider {
+    /// Creates a new time provider
+    pub fn new(time_getter: impl GetCurrentTime + 'static) -> Self {
+        Self {
+            inner: Arc::new(time_getter),
+        }
+    }
+
+    /// A time provider that always fails
+    pub fn none() -> Self {
```
Comment (djc) [inline rustls/src/time_provider.rs:26]: How does it make sense to have this? Is it possible to do TLS without time?

File: rustls/src/time_provider.rs [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+use core::fmt::Debug;
+
+use pki_types::UnixTime;
+
+/// An object that provides the current time.
+///
+/// This is used to, for example, check if a certificate has expired during certificate validation,
+/// or to check the age of a ticket.
+#[derive(Clone, Debug)]
+pub struct TimeProvider {
```
Comment (djc) [inline rustls/src/time_provider.rs:13]: What value does this wrapper provide? IMO it would be more clear to avoid this wrapper and use `Arc<dyn GetCurrentTime>` directly.

File: rustls/src/time_provider.rs [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+        Self::new(NoTimeProvider)
+    }
+
+    pub(crate) fn get_current_time(&self) -> Option<UnixTime> {
+        self.inner.get_current_time()
+    }
+}
+
+/// Get current time
+pub trait GetCurrentTime: Debug + Send + Sync {
```
Comment (djc) [inline rustls/src/time_provider.rs:36]: Not a big fan of this trait name, I'd name this `TimeProvider`.

File: rustls/src/msgs/persist.rs [hunk 14]
```diff
@@ -406,9 +406,11 @@ impl ServerSessionValue {
 #[cfg(test)]
 mod tests {
     use super::*;
+    #[cfg(feature = "std")]
```
Comment (djc) [inline rustls/src/msgs/persist.rs:409]: Why is this necessary? Seems surprising.

File: rustls/src/server/server_conn.rs [hunk 15]
```diff
@@ -405,6 +411,19 @@ impl ServerConfig {
             .iter()
             .any(|cs| cs.usable_for_protocol(proto))
     }
+
+    pub(super) fn get_current_time(&self) -> Result<UnixTime, Error> {
```
Comment (djc) [inline rustls/src/server/server_conn.rs:422]: Name, as for `ClientConfig`.

File: rustls/src/lib.rs [hunk 16]
```diff
@@ -470,7 +470,14 @@ use crate::crypto::ring as test_provider;
 // The public interface is:
 pub use crate::builder::{ConfigBuilder, ConfigSide, WantsVerifier, WantsVersions};
 pub use crate::common_state::{CommonState, IoState, Side};
-pub use crate::conn::{Connection, ConnectionCommon, Reader, SideData, Writer};
+pub use crate::conn::unbuffered::{
```
Comment (djc) [inline rustls/src/lib.rs:474]: Not sure this change makes sense in this commit?

File: rustls/Cargo.toml [hunk 17]
```diff
@@ -29,7 +29,7 @@ default = ["logging", "ring", "tls12", "std"]
 std = ["webpki/std", "pki-types/std"]
 logging = ["log"]
 aws_lc_rs = ["dep:aws-lc-rs", "webpki/aws_lc_rs", "std"]
-ring = ["dep:ring", "webpki/ring", "std"]
+ring = ["dep:ring", "webpki/ring"]
```
Comment (djc) [inline rustls/Cargo.toml:32]: IMO this should be squashed into the change at the start, so we're not flipflopping within the PR.

**Review by djc — COMMENTED** [review]
It's unclear to me how usable `no_std` rustls is after this PR.

Does this pass `cargo check` without warnings at every commit? What about `cargo check --all-targets`?

The proliferation of `std` guards seems pretty ugly, it might be better if there was some grouping with inline modules.

File: rustls/src/time_provider.rs (hunk 11 above)
Comment (pvdrz) [reply rustls/src/time_provider.rs:26]: As mentioned in [the RFC](https://github.com/japaric/rustls/blob/ja-no-std-support-rfc/rfcs/000-no-std-support.md#systemtimenow) this is the default time provider and will always error as the user must provide their own time provider in no-std contexts.

Comment (djc) [reply rustls/src/time_provider.rs]: But why do we want to provide a default time provider that doesn't actually provide time? Is it load-bearing that time providers can yield `None`?

**Comment by pvdrz** [comment]
> It's unclear to me how usable `no_std` rustls is after this PR.

As stated in the [RFC](https://github.com/japaric/rustls/blob/ja-no-std-support-rfc/rfcs/000-no-std-support.md#first-stage), the unbuffered API from the client side should be usable (without support sessions)
> Does this pass `cargo check` without warnings at every commit? What about `cargo check --all-targets`?

Yes, both `cargo check` and `cargo check --all-targets` pass.

> The proliferation of `std` guards seems pretty ugly, it might be better if there was some grouping with inline modules.

I'll see what I can do.

**Comment by pvdrz** [comment]
@djc I moved most of the std only items to inline modules or to gated `impl` blocks. The ones remaining weren't worth the change imho. Some of these changes will be reverted by https://github.com/rustls/rustls/pull/1688 so I don't think they will be a big deal and such adjustments can be done on that PR later

File: rustls/src/error.rs [hunk 18]
```diff
@@ -76,7 +76,7 @@ pub enum Error {
     General(String),
 
     /// We failed to figure out what time it currently is.
-    FailedToGetCurrentTime,
+    FailedToTimeProvider,
```
Comment (djc) [inline rustls/src/error.rs:78]: Guessing this is some kind of mass-rename fallout? 🙂 I'd expect the error variants to stick to `FailedToGetCurrentTime`.

File: rustls/src/client/hs.rs [hunk 19]
```diff
@@ -68,7 +68,12 @@ fn find_session(
             None
         })
         .and_then(|resuming| {
-            let retrieved = persist::Retrieved::new(resuming, UnixTime::now());
+            let now = config
+                .current_time()
+                .map_err(|_err| debug!("Could not get current time: {}", _err))
```
Comment (djc) [inline rustls/src/client/hs.rs:73]: Style nit: please inline format arguments.

File: rustls/src/lib.rs (hunk 16 above)
Comment (djc) [reply rustls/src/lib.rs:474]: Okay, at least they're in a separate commit now, but I'm still confused. We already expose all of this stuff in the `unbuffered` module, around line 454?

File: rustls/src/server/server_conn.rs [hunk 20]
```diff
@@ -418,160 +417,178 @@ impl ServerConfig {
...
-/// Allows reading of early data in resumed TLS1.3 connections.
-///
-/// "Early data" is also known as "0-RTT data".
-///
-/// This structure implements [`std::io::Read`].
-pub struct ReadEarlyData<'a> {
-    early_data: &'a mut EarlyDataState,
-}
+#[cfg(feature = "std")]
+mod server_connection {
```
Comment (djc) [inline rustls/src/server/server_conn.rs:425]: Let's just call this `connection`? (Same on the client side.)

File: rustls/src/common_state.rs [hunk 21]
```diff
@@ -51,6 +51,66 @@ pub struct CommonState {
     pub(crate) enable_secret_extraction: bool,
 }
 
+#[cfg(feature = "std")]
+impl CommonState {
```
Comment (djc) [inline rustls/src/common_state.rs:56]: Style nit: I like the idea of splitting the `impl`, but I think the guarded `impl` should go directly below the unguarded one.

(Same for other split `impl`s.)

File: rustls/src/server/handy.rs [hunk 22]
```diff
@@ -141,71 +135,114 @@ impl server::ResolvesServerCert for AlwaysResolvesChain {
...
 }
 
-/// Something that resolves do different cert chains/keys based
-/// on client-supplied server name (via SNI).
-#[derive(Debug)]
-pub struct ResolvesServerCertUsingSni {
-    by_name: HashMap<String, Arc<sign::CertifiedKey>>,
-}
+#[cfg(feature = "std")]
+mod resolves_server_cert_using_sni {
```
Comment (djc) [inline rustls/src/server/handy.rs:197]: Nit: suggest naming this `sni_resolver`.

File: rustls/src/server/handy.rs [hunk 23]
```diff
@@ -28,56 +26,116 @@ impl server::StoresServerSessions for NoServerSessionStorage {
...
 }
 
-/// An implementer of `StoresServerSessions` that stores everything
-/// in memory.  If enforces a limit on the number of stored sessions
-/// to bound memory usage.
-pub struct ServerSessionMemoryCache {
-    cache: Mutex<limited_cache::LimitedCache<Vec<u8>, Vec<u8>>>,
-}
+#[cfg(feature = "std")]
+mod server_session_memory_cache {
```
Comment (djc) [inline rustls/src/server/handy.rs:30]: Nit: suggest naming this `cache`.

**Review by djc — COMMENTED** [review]
Okay, this is much better -- I think the main sticking point is the `TimeProvider` API.

**Comment by djc** [comment]
> @djc I moved most of the std only items to inline modules or to gated `impl` blocks. The ones remaining weren't worth the change imho. Some of these changes will be reverted by #1688 so I don't think they will be a big deal and such adjustments can be done on that PR later

Yup, this is much better, thanks for working on that!

File: rustls/src/time_provider.rs (hunk 11 above)
Comment (djc) [reply rustls/src/time_provider.rs]: So as far as I can surmise, the only reason to have `TimeProvider::current_time()` yield `Option` is that the config builder must create a `TImeProvider` to build a complete config, and in `no_std` situation that is platform dependent so we can't create it. I think we should probably force `no_std` configurations through a new separate `ClientConfig::builder_with_time_provider()` path instead and make `TimeProvider` yield just `UnixTime` or, more likely, `Result<UnixTime, Error>`.

Comment (pvdrz) [reply rustls/src/time_provider.rs]: some real-time clock drivers are fallible [ds1307](https://docs.rs/ds1307/latest/ds1307/#set-and-get-the-current-date-and-time-all-at-once) is an example of this. Returning `Option<UnixTime>` allows representing such errors, you could argue that we could return `Result<UnixTime, TimeProvider::Error>` or something like that but I consider that the error provides little to no use.

File: rustls/src/error.rs (hunk 18 above)
Comment (pvdrz) [reply rustls/src/error.rs:78]: Weird, I did that rename due to a rebase conflict. Maybe I'm confused but I'll revert it as the original PR used `FailedToGetCurrentTime` as well.

**Comment by cpu** [comment]
@ctz @djc I've rebased this branch on top of `main` and I think it's ready for review. This branch is mostly feature-gating things that are `std`-only and then adding a `x86_64-unknown-none` build to keep us from backsliding so it might be nice to land this soon so we don't accidentally trip into choices that will make things harder in the future.

Most of my changes were pretty boring (_resolving conflicts, fixing some new `clippy` findings, etc_). The one place worth an explicit note is that I had to do some fiddling in the first commit (ae1c3a5744e57db54890220835e18d6491171e1c) in order to get the no-std build working with the `once_cell` dependency that landed in-tree in e7a1b418527eceba639457dc2ad2dcb29078f612 between the time of the original work by Japaric  et. al. and now.

The `once_cell` crate has a [FAQ section](https://docs.rs/once_cell/latest/once_cell/#faq) titled "Does this crate support no_std?" where it offers two potential solutions. I went with the `critical-section` feature but I'm out of my depth here. It seems like there's some nuance with respect to the right way to handle this and while my choice seems to appease the `x86_64-unknown-none` target build in CI and on my local machine I'm not sure if it's truly the best approach. WDYT?

**Comment by cpu** [comment]
Rebased on `main`

**Comment by nspin** [comment]
> The once_cell crate has a [FAQ section](https://docs.rs/once_cell/latest/once_cell/#faq) titled "Does this crate support no_std?" where it offers two potential solutions. I went with the critical-section feature but I'm out of my depth here. It seems like there's some nuance with respect to the right way to handle this and while my choice seems to appease the x86_64-unknown-none target build in CI and on my local machine I'm not sure if it's truly the best approach. WDYT?

In my opinion, the dependency on `critical-section` is unnecessary, and using [`once_cell::race::OnceBox`](https://docs.rs/once_cell/latest/once_cell/race/struct.OnceBox.html) would suffice.

The only cost of [`OnceBox`](https://docs.rs/once_cell/latest/once_cell/race/struct.OnceBox.html) that would be relevant to its usage in Rustls is that it introduces another level of pointer indirection, but `PROCESS_DEFAULT_PROVIDER` isn't used in any performance-critical paths, so I don't see this as a problem.

Blocking (via `std` or `critical-section`) provides better semantics for the [`get_or_init`](https://docs.rs/once_cell/latest/once_cell/sync/struct.OnceCell.html#method.get_or_init) family of functions, but Rustls does not use these.

**Comment by cpu** [comment]
> In my opinion, the dependency on critical-section is unnecessary, and using [once_cell::race::OnceBox](https://docs.rs/once_cell/latest/once_cell/race/struct.OnceBox.html) would suffice.

@nspin Thanks for your input, much appreciated :bow: I've switched over to using `race` and `OnceBox` in my last push alongside rebasing on `main`.

File: quic.fail.txt [hunk 24]
```diff
@@ -0,0 +1,124 @@
+an_send_3_inital_packets stdout ----
```
[the remark this answers is not in the document]
Comment (cpu) [reply quic.fail.txt:1]: :sweat_smile: Yup, my bad. Thanks for noticing.

[the remark this answers is not in the document]
Comment (cpu) [reply quic.fail.txt]: Fixed. I'll work on the other feedback shortly.

File: rustls/src/lib.rs [hunk 25]
```diff
@@ -610,3 +610,8 @@ pub mod ticketer;
 
 /// This is the rustls manual.
 pub mod manual;
+
+#[cfg(feature = "std")]
+mod time_provider;
+#[cfg(not(feature = "std"))]
+pub mod time_provider;
```
[the remark this answers is not in the document]
Comment (cpu) [reply rustls/src/lib.rs]: Switched to have this module public both with and without `std`.

File: rustls/src/error.rs [hunk 26]
```diff
@@ -544,43 +546,56 @@ impl From<rand::GetRandomFailed> for Error {
...
 
-/// Any other error that cannot be expressed by a more specific [`Error`] variant.
-///
-/// For example, an `OtherError` could be produced by a custom crypto provider
-/// exposing a provider specific error.
-///
-/// Enums holding this type will never compare equal to each other.
-#[derive(Debug, Clone)]
-pub struct OtherError(pub Arc<dyn StdError + Send + Sync>);
+#[cfg(feature = "std")]
```
[the remark this answers is not in the document]
Comment (cpu) [reply rustls/src/error.rs]: I've reworked the relevant commit. Good suggestion, it seems cleaner to me :+1:

File: rustls/src/conn.rs [hunk 27]
```diff
@@ -372,9 +373,127 @@ pub struct ConnectionCommon<Data> {
...
+    /// Success from this function comes with some sundry state data
+    /// about the connection.
+    ///
+    /// [`read_tls`]: Connection::read_tls
+    /// [`process_new_packets`]: Connection::process_new_packets
+    #[inline]
+    pub fn process_new_packets(&mut self) -> Result<IoState, Error> {
+        self.core
+            .process_new_packets(&mut self.deframer_buffer, &mut self.sendable_plaintext)
+    }
```
[the remark this answers is not in the document]
Comment (cpu) [reply rustls/src/conn.rs:399]: Fixed. Thanks

File: rustls/src/server/builder.rs [hunk 28]
```diff
@@ -116,7 +116,10 @@ impl ConfigBuilder<ServerConfig, WantsServerCert> {
             cert_resolver,
             ignore_client_order: false,
             max_fragment_size: None,
+            #[cfg(feature = "std")]
             session_storage: handy::ServerSessionMemoryCache::new(256),
+            #[cfg(not(feature = "std"))]
+            session_storage: Arc::new(handy::NoServerSessionStorage {}),
```
[the remark this answers is not in the document]
Comment (cpu) [reply rustls/src/server/builder.rs:122]: I've updated the docs. We'll also need to remember to update them again when https://github.com/rustls/rustls/pull/1688 re-introduces session storage for no-std w/ hashbrown. :reminder_ribbon: :ballot_box_with_check:

File: rustls/src/error.rs [hunk 29]
```diff
@@ -544,43 +546,59 @@ impl From<rand::GetRandomFailed> for Error {
...
     }
-}
 
-impl fmt::Display for OtherError {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}", self.0)
+    #[cfg(feature = "std")]
+    impl fmt::Display for OtherError {
+        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            write!(f, "{}", self.0)
```
Comment (ctz) [inline rustls/src/error.rs:585]:
```suggestion
    impl fmt::Display for OtherError {
        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
            #[cfg(feature = "std")]
            write!(f, "{}", self.0)
            #[cfg(not(feature = "std"))]
            f.write_str("no further information available")
```

I think that should address the remaining gating in `impl fmt::Display for Error` above.

**Review by ctz — APPROVED** [review]

File: rustls/src/time_provider.rs [hunk 30]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+#[derive(Debug)]
+pub(crate) struct DefaultTimeProvider;
+
+impl TimeProvider for DefaultTimeProvider {
+    fn current_time(&self) -> Option<UnixTime> {
+        #[cfg(feature = "std")]
+        return Some(UnixTime::now());
+        #[cfg(not(feature = "std"))]
```
Comment (djc) [inline rustls/src/time_provider.rs:26]: As explained [previously](https://github.com/rustls/rustls/pull/1502#discussion_r1434942346), I'm not a fan of this setup. Can we instead feature-gate the `DefaultTimeProvider` on `#[cfg(feature = "std")]`, forcing `no_std` users to supply their own?

File: rustls/src/time_provider.rs [hunk 31]
```diff
@@ -0,0 +1,29 @@
...
+
+use core::fmt::Debug;
+
+use pki_types::UnixTime;
+
+/// An object that provides the current time.
+///
+/// This is used to, for example, check if a certificate has expired during certificate validation,
+/// or to check the age of a ticket.
+#[allow(unreachable_pub)]
```
Comment (djc) [inline rustls/src/time_provider.rs:11]: Not sure this `unreachable_pub` makes sense, we should allow access to this before this PR is merged, right -- so might as well do it in this commit?

Comment (djc) [reply rustls/src/time_provider.rs:11]: (Maybe squash this commit with the first one that actually introduces usage of the trait?)

File: rustls/src/time_provider.rs (hunk 30 above)
Comment (djc) [reply rustls/src/time_provider.rs:26]: Also, if every caller of `current_time()` is going to convert the `Option` to `Result` anyway, should we just allow the implementer of this trait to supply an `Error` directly?

Comment (cpu) [reply rustls/src/time_provider.rs:26]: > Can we instead feature-gate the DefaultTimeProvider on #[cfg(feature = "std")], forcing no_std users to supply their own?

I can give this a shot. @ctz Do you prefer this approach as well?

> I suppose the no_std users who are the most likely implementers don't currently have a Box<dyn std::error::Error> to smuggle an alternative error through...

Yeah, that seems like a good reason to leave it as an `Option` IMO.

File: rustls/src/time_provider.rs (hunk 31 above)
Comment (cpu) [reply rustls/src/time_provider.rs]: Fixed.

**Comment by cpu** [comment]
> cpu force-pushed the no-std-support branch from 4251a12 to 231cfdf 

Rebased to address conflicts w/ `main`.

File: rustls/src/client/builder.rs [hunk 32]
```diff
@@ -161,6 +166,97 @@ impl ConfigBuilder<ClientConfig, WantsClientCert> {
...
+#[cfg(not(feature = "std"))]
+#[derive(Clone)]
+pub struct WantsTimeProvider {
+    provider: Arc<CryptoProvider>,
+    versions: versions::EnabledVersions,
+    verifier: Arc<dyn verify::ServerCertVerifier>,
+    client_auth_cert_resolver: Arc<dyn ResolvesClientCert>,
+}
+
+#[cfg(not(feature = "std"))]
```
[the remark this answers is not in the document]
Comment (cpu) [reply rustls/src/client/builder.rs:238]: Would the appeal be maintaining symmetry, or is there a deeper usecase you had in mind where you would want to provide a time provider w/ `std` enabled?

I don't have a strong preference here so would generally favour whatever Djc thinks is best since his feedback prompted the initial change from the approach used by the PR's first author.

[the remark this answers is not in the document]
Comment (cpu) [reply rustls/src/client/builder.rs:238]: > Sorry for the drive-by comment

Happy for input here btw :-)

[the remark this answers is not in the document]
Comment (ctz) [reply rustls/src/client/builder.rs:238]: It feels less hazardous (in the sense of crate features changing return types, etc.) to:

- std-gate all existing `ClientConfig::builder*`
- add ungated `ClientConfig::builder_with_details` (or some better name) that takes absolutely everything: a CryptoProvider, this time provider, (potentially a `Resumption` to put the no-std unhelpful and surprising `Resumption::default` front-and-center?)

[the remark this answers is not in the document]
Comment (djc) [reply rustls/src/client/builder.rs:238]: Yeah, hiding API when a feature is enabled is generally a no-no because Cargo features are supposed to be additive.

[the remark this answers is not in the document]
Comment (ctz) [reply rustls/src/client/builder.rs:238]: To clarify: do you mean adding a parameter to `builder_with_provider`? Or is a new `builder_with_*` function OK?

(I think my opinion is: 100% of `builder_with_provider` callers are std currently, and inflicting that change on them seems a bit gratuitous; so I prefer a new API surface.)

[the remark this answers is not in the document]
Comment (ctz) [reply rustls/src/client/builder.rs]: I have pushed a commit (that can be dissolved into earlier ones) reflecting my preferences on this.

[the remark this answers is not in the document]
Comment (djc) [reply rustls/src/client/builder.rs]: Yup, that's pretty much exactly what I had in mind. Tried to explain (badly) in https://github.com/rustls/rustls/pull/1502#discussion_r1434942346.

**Review by djc — APPROVED** [review]

File: rustls/src/client/builder.rs (hunk 32 above)
[the remark this answers is not in the document]
Comment (ctz) [reply rustls/src/client/builder.rs]: I've squashed this into the commits that introduced it. I also took the liberty of improving some of the docs around `TimeProvider`

[the remark this answers is not in the document]
Comment (cpu) [reply rustls/src/client/builder.rs]: This direction makes sense to me :+1:

**Review by cpu — APPROVED** [review]
Thanks for landing those last tweaks @ctz
